### PR TITLE
Fix NSUInteger size for Watchos x64

### PR DESCRIPTION
**Pull request by PavelPunegov**
Fix for the https://youtrack.jetbrains.com/issue/KT-44083

### CODE REVIEW

File: backend.native/tests/interop/objc/smoke.kt [hunk 1]
```diff
@@ -67,11 +67,12 @@ fun run() {
 
     // hashCode (directly):
     // hash() returns value of NSUInteger type.
-    val hash = when (Platform.osFamily) {
+    val hash = if (Platform.osFamily == OsFamily.WATCHOS && Platform.cpuArchitecture.bitness == 32) {
```
Comment (sbogolepov) [inline backend.native/tests/interop/objc/smoke.kt:70]: This check is not correct for `watchos_arm64` (its bitness is 64 and NSInteger is defined as int). Though we are not testing this target, this piece of code might be confusing for readers.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt [hunk 2]
```diff
@@ -1563,34 +1563,22 @@ internal fun ObjCExportCodeGenerator.getEncoding(methodBridge: MethodBridge): St
...
-    Family.WATCHOS -> "I"
-    else -> error("Unexpected target platform: $this")
-}
-
-private fun MethodBridge.ReturnValue.getObjCEncoding(targetFamily: Family): String = when (this) {
+private fun MethodBridge.ReturnValue.getObjCEncoding(context: Context): String = when (this) {
     MethodBridge.ReturnValue.Suspend,
     MethodBridge.ReturnValue.Void -> "v"
-    MethodBridge.ReturnValue.HashCode -> targetFamily.nsUIntegerEncoding
+    MethodBridge.ReturnValue.HashCode -> if (context.is64BitNSInteger()) "L" else "I"
```
Comment (sbogolepov) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1575]: Did you check it for `ios_arm32`?

Comment (PavelPunegov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1575]: We don't have any devices that support that target

File: backend.native/tests/interop/objc/smoke.kt (hunk 1 above)
Comment (PavelPunegov) [reply backend.native/tests/interop/objc/smoke.kt]: Changed the check

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt (hunk 2 above)
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1575]: I'm not sure this is entirely correct. `L` doesn't mean "64-bit intenger", it is used for C `unsigned long` type. So `NSUInteger` should still be `L` on ios_arm32. See also https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html

File: backend.native/tests/interop/objc/smoke.kt [hunk 3]
```diff
@@ -67,11 +67,12 @@ fun run() {
 
     // hashCode (directly):
     // hash() returns value of NSUInteger type.
-    val hash = when (Platform.osFamily) {
+    val hash = if (Platform.osFamily == OsFamily.WATCHOS && Platform.cpuArchitecture != CpuArchitecture.X64) {
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/interop/objc/smoke.kt:70]: Checking `sizeOf<NSUInteger>() == 4` instead also might make sense.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt (hunk 2 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1575]: JFTR: we've rechecked and apparently things are even more complicated here:
https://github.com/llvm/llvm-project/blob/c161775decddcc86fbbfefd7485a5d0ef5842aec/clang/lib/AST/ASTContext.cpp#L7163
(which kinda doesn't match the document referred above).

Current PR version still might be incorrect, probably needs more work.

Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1575]: Yes, but we can compare compilers output. E.g. check what Clang does.

File: backend.native/tests/interop/objc/smoke.kt [hunk 4]
```diff
@@ -67,11 +67,12 @@ fun run() {
 
     // hashCode (directly):
     // hash() returns value of NSUInteger type.
-    val hash = when (Platform.osFamily) {
+    val hash = if (sizeOf<NSUIntegerVar>() == 4L) {
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/interop/objc/smoke.kt:70]: The comments below are obsolete too.

**Review by SvyatoslavScherbina — APPROVED** [review]

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt [hunk 5]
```diff
@@ -1563,34 +1563,22 @@ internal fun ObjCExportCodeGenerator.getEncoding(methodBridge: MethodBridge): St
...
-    Family.WATCHOS -> "I"
-    else -> error("Unexpected target platform: $this")
-}
-
-private fun MethodBridge.ReturnValue.getObjCEncoding(targetFamily: Family): String = when (this) {
+private fun MethodBridge.ReturnValue.getObjCEncoding(context: Context): String = when (this) {
     MethodBridge.ReturnValue.Suspend,
     MethodBridge.ReturnValue.Void -> "v"
-    MethodBridge.ReturnValue.HashCode -> targetFamily.nsUIntegerEncoding
+    MethodBridge.ReturnValue.HashCode -> if (context.is64BitNSInteger()) "Q" else "I"
```
Comment (sbogolepov) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1575]: IIUC, It should be "L" for ios_arm32:
* `NSUInteger` is an alias for `unsigned long` on iOS.
* `unsigned long` is mapped to "L" on 32-bit platforms.
Am I missing something?

Comment (PavelPunegov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1575]: See [comment in the issue](https://youtrack.jetbrains.com/issue/KT-44083#focus=Comments-27-4644443.0-0) with details of the `clang` output. It shows that it encodes to `I` on 32-bit iOS.

**Review by sbogolepov — APPROVED** [review]

**Pull request merged**